### PR TITLE
Fixes array_intersect_key error when the 1st argument is a string

### PR DESCRIPTION
**Pull request by aristath**
Trac ticket: https://core.trac.wordpress.org/ticket/51986

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
@aristath thanks for the patch! I think I would prefer if we moved this to `register_rest_route` like we do for our other `_doing_it_wrong` notices. It should hopefully make it more clear about where the error is coming from and we can make sure it is only printed once per route. Something like

```
if ( count( array_filter( $args, 'is_array' ) ) !== count( $args ) {

}
```

**Comment by aristath** [comment]
Thank you @TimothyBJacobs,
I moved the `_doing_it_wrong` call to `register_rest_route`. I did however leave the extra check in `get_data_for_route` as that will act as a safeguard in case a REST route is registered via `rest_get_server()->register_route` or other methods and not using the `register_rest_route` function.

**Comment by TimothyBJacobs** [comment]
Thanks for updating this @aristath! We should be checking `$arg_group['args']` not `$args` though. Let's also make sure to include the namespace and route in the doing it wrong message so it will be cleared to the developers exactly which route registration is failing.

We can also make the check in `get_data_for_route` use `is_array` as well since any non-array value will cause PHP warnings, not just strings.

Lastly, would you be interested in writing unit tests for this? You can see `\Tests_REST_API::test_register_route_with_missing_permission_callback_top_level_route` for some examples of how to test `register_rest_route` with an invalid definition.

File: src/wp-includes/rest-api/class-wp-rest-server.php [hunk 1]
```diff
@@ -1399,6 +1399,11 @@ public function get_data_for_route( $route, $callbacks, $context = 'view' ) {
 				$endpoint_data['args'] = array();
 
 				foreach ( $callback['args'] as $key => $opts ) {
+					if ( is_string( $opts ) ) {
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/class-wp-rest-server.php:1402]: Do we need this check here? It looks like the value will just end up getting stripped out by the `array_intersect_key` check, right?

**Comment by TimothyBJacobs** [comment]
Hey @aristath,

Just checking in about unit tests.

**Comment by aristath** [comment]
@TimothyBJacobs I tried but I'm afraid I don't understand how to write unit tests for that.
If you want I can just close this PR so you can create another one with tests included.

**Comment by nate-allen** [comment]
I [added a unit test](https://core.trac.wordpress.org/attachment/ticket/51986/51986.test.diff) based on the code from this PR .

**Comment by TimothyBJacobs** [comment]
Thanks @nate-allen! Could you create a new PR with @aristath's changes plus your unit test?

**Comment by dream-encode** [comment]
Merged into core in https://core.trac.wordpress.org/changeset/54339.

**Pull request closed without merging**